### PR TITLE
Record job start time to fix time pain points

### LINKED ISSUE #1544: Epoch issues
**Please acknowledge the following before creating a ticket**

- [X] I have read the GitHub issues section of [REPORTING-BUGS](../blob/master/REPORTING-BUGS).

**Description of the bug:**
By default, fio log files use timestamps based on an epoch that is the beginning of the job. If you have multiple jobs, especially with some jobs waiting on other jobs before starting, this means you have different epochs and the times in the logs for each job cannot be compared/ordered.

Using `log_alternate_epoch` solves this problem as the logs all then share a common epoch. Unfortunately, it introduces a new problem. That is, the start time for each job relative to this common epoch is not known. So the gap between a job starting and the first log entry is indeterminate (although typically small, 0-1ms). (Less critically, `log_alternate_epoch` also increases log file size because the number of characters for each log entry timestamp increases.)

The shortcomings of each approach can be resolved by using the beginning of job epoch for log files but recording the start time of each job against a common (alternate) epoch.

I put something like this together in this PR from a while back: https://github.com/axboe/fio/pull/1353/commits

Back then I was more concerned about correlating log entries against other events on the same machine, and the PR was aimed at simplifying log files while still being able to know time against a machine-based epoch. It was nice to have but not super important (imho).

But as I've started to work with fio experiments with multiple jobs, having timekeeping shortcomings both with the normal epoch and `log_alternate_epoch` has led me to the same solution (the PR) to a different problem.

I was hoping the PR could be reconsidered or maybe some other solution brainstormed? I'm happy to implement a solution, whether it looks like the PR or something completely different.
**Environment**: 
All

**fio version**:
All

**Reproduction steps**
A simple fio run with per sample logging, each job waiting on the previous, illustrates the issues

### PR DESCRIPTION
**Pull request by aggieNick02**
Add a new key in the json per-job output, job_start, that records the job start time obtained via a call to clock_gettime using the clock_id specified by the new job_start_clock_id option. This allows times of fio jobs and log entries to be compared/ordered against each other and against other system events recorded against the same clock_id.

Also make log_unix_epoch an official alias of log_alternate_epoch, instead of maintaining both redundant options.

Add a note to the documentation for group_reporting about how there are several per-job values for which only the first job's value is recorded in the json output format when group_reporting is enabled.

Fixes #1544

Signed-off-by: Nick Neumann nick@pcpartpicker.com

### CODE REVIEW

**Comment by aggieNick02** [comment]
Unrelated to the content of the PR - in testing client/server, I spun my wheels for a bit before I realized that command-line options from the client don't really seem to work - instead, everything needs to be in a job file. Any idea if this is known/expected? The docs make it look like command-line options are supported, but it doesn't seem like it. Wondering if it's a bug or if the docs need fixing.

File: client.c [hunk 1]
```diff
@@ -956,6 +956,7 @@ static void convert_ts(struct thread_stat *dst, struct thread_stat *src)
 	dst->error		= le32_to_cpu(src->error);
 	dst->thread_number	= le32_to_cpu(src->thread_number);
 	dst->groupid		= le32_to_cpu(src->groupid);
+	dst->job_start = le64_to_cpu(src->job_start);
```
Comment (vincentkfu) [inline client.c:959]: Can you line this up like the other lines?

File: server.c [hunk 2]
```diff
@@ -1706,6 +1706,7 @@ void fio_server_send_ts(struct thread_stat *ts, struct group_run_stats *rs)
 	p.ts.error		= cpu_to_le32(ts->error);
 	p.ts.thread_number	= cpu_to_le32(ts->thread_number);
 	p.ts.groupid		= cpu_to_le32(ts->groupid);
+	p.ts.job_start = cpu_to_le64(ts->job_start);
```
Comment (vincentkfu) [inline server.c:1709]: Please line this up like the other lines

**Comment by vincentkfu** [comment]
The code changes all look reasonable. The only issue I see is that there is a single patch doing two things (adding the new key and making two options synonyms of each other). Could you separate these into two distinct patches?

**Comment by vincentkfu** [comment]
> Unrelated to the content of the PR - in testing client/server, I spun my wheels for a bit before I realized that command-line options from the client don't really seem to work - instead, everything needs to be in a job file. Any idea if this is known/expected? The docs make it look like command-line options are supported, but it doesn't seem like it. Wondering if it's a bug or if the docs need fixing.

This sounds like something we should fix, although it's not clear whether we should change the documentation to match fio's behavior or vice versa.

File: client.c (hunk 1 above)
Comment (aggieNick02) [reply client.c]: Sure, sorry about that.

File: server.c (hunk 2 above)
Comment (aggieNick02) [reply server.c]: Sure, sorry about that.

**Comment by aggieNick02** [comment]
> The code changes all look reasonable. The only issue I see is that there is a single patch doing two things (adding the new key and making two options synonyms of each other). Could you separate these into two distinct patches?

Sure, I can definitely do this if you want. The one wrinkle in doing so is that there will be conflicts as the changes touch some of the same lines. Do you want me to first do the new key and then wait until that PR is merged and then do the synonyms?

**Comment by aggieNick02** [comment]
> > Unrelated to the content of the PR - in testing client/server, I spun my wheels for a bit before I realized that command-line options from the client don't really seem to work - instead, everything needs to be in a job file. Any idea if this is known/expected? The docs make it look like command-line options are supported, but it doesn't seem like it. Wondering if it's a bug or if the docs need fixing.
> 
> This sounds like something we should fix, although it's not clear whether we should change the documentation to match fio's behavior or vice versa.

I'll take a quick look and see if I can see what might be going wrong.

**Comment by vincentkfu** [comment]
> > The code changes all look reasonable. The only issue I see is that there is a single patch doing two things (adding the new key and making two options synonyms of each other). Could you separate these into two distinct patches?
> 
> Sure, I can definitely do this if you want. The one wrinkle in doing so is that there will be conflicts as the changes touch some of the same lines. Do you want me to first do the new key and then wait until that PR is merged and then do the synonyms?

I would just have this one PR include both patches. Either order is fine.

**Comment by aggieNick02** [comment]
> I would just have this one PR include both patches. Either order is fine.

Ah, ok. I just updated the PR with two separate commits. Thanks!

**Pull request merged**